test/unit/org/apache/cassandra/cql3/validation/operations/CQLVectorTest.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.cql3.validation.operations;
+
+import org.junit.Test;
+
+import org.apache.cassandra.cql3.CQLTester;
+
+public class CQLVectorTest extends CQLTester.InMemory
```

Comment:
I think we should probably add some tests for functions. For example, the following doesn't work:
```java
@Test
public void functions()
{
    VectorType<Integer> type = VectorType.getInstance(Int32Type.instance, 2);
    Vector<Integer> vector = vector(1, 2);

    NativeFunctions.instance.add(new NativeScalarFunction("f", type, type)
    {
        @Override
        public ByteBuffer execute(ProtocolVersion protocol, List<ByteBuffer> parameters)
        {
            return parameters.get(0);
        }
    });
    
    createTable(KEYSPACE, "CREATE TABLE %s (pk int primary key, value vector<int, 2>)");
    execute("INSERT INTO %s (pk, value) VALUES (0, ?)", vector);

    assertRows(execute("SELECT f(value) FROM %s WHERE pk=0"), row(vector));
    assertRows(execute("SELECT f([1, 2]) FROM %s WHERE pk=0"), row(vector));
}
```
It throws `Type error: [1, 2] cannot be passed as argument 0 of function system.f of type vector<int, 2>`. 

That's because the terminal argument in the second function call is a `Selectable.WithList`, and that selectable refuses to be assigned to the vector type. That happens in `Selectable.WithList#testAssignment`.

Reply:
This fails at `assertRows(execute("SELECT f([1, 2]) FROM %s WHERE pk=0"), row(vector));`. In a debugger `[1, 2]` is a `ListType` and not `ArrayLiteral`, so can't convert `ListType` to `VectorType`

Not actually sure why this isn't `ArrayLiteral`; ill debug

Reply:
ignore me, the list is a list of literals and the literal is 

```
selectable = {Selectable$WithList@7024} "[1, 2]"
 selectables = {ArrayList@8411}  size = 2
  0 = {Selectable$WithTerm@8413} "1"
  1 = {Selectable$WithTerm@8414} "2"
```

so isn't `ArrayLiteral` but its own thing

Reply:
ok found it, `org.apache.cassandra.cql3.Lists#testListAssignment`

```
if (!(receiver.type instanceof ListType))
    return AssignmentTestable.TestResult.NOT_ASSIGNABLE;
```

Reply:
ok I pushed a change with this test but commented out `assertRows(execute("SELECT f([1, 2]) FROM %s WHERE pk=0"), row(vector));`.  The issue is `Selectable.WithList` doesn't understand `VectorType` and only works with `ListType`, so we need the same trick as `ArrayLiteral` to proxy once we know the target...

I choose not to fix this logic as this patch does not add functions, it only adds the type... so I will leave fixing `Selectables` to w/e adds functions... I was nice enough to add everything needed in `Vectors` class you would need to impl

Reply:
The problem with vector terminals doesn't need any new functions to reproduce. It also breaks existing functions like the venerable `token` function:
```java
@Test
public void token()
{
    createTable(KEYSPACE, "CREATE TABLE %s (pk vector<int, 2> primary key)");
    execute("INSERT INTO %s (pk) VALUES (?)", vector(1, 2));
    long tokenColumn = execute("SELECT token(pk) as t FROM %s").one().getLong("t");
    long tokenTerminal = execute("SELECT token([1, 2]) as t FROM %s").one().getLong("t"); // Type error
    Assert.assertEquals(tokenColumn, tokenTerminal);
}
```
Also, users can try to use vectors on their UDFs, where they will hit the same problem:
```java
@Test
public void udf() throws Throwable
{
    createTable(KEYSPACE, "CREATE TABLE %s (pk int primary key, value vector<int, 2>)");
    String function = createFunction(KEYSPACE,
                                     "",
                                     "CREATE FUNCTION %s (x vector<int, 2>) " +
                                     "CALLED ON NULL INPUT " +
                                     "RETURNS vector<int, 2> " +
                                     "LANGUAGE java " +
                                     "AS 'return x;'");

    Vector<Integer> value = vector(1, 2);
    execute("INSERT INTO %s (pk, value) VALUES (0, ?)", value);

    assertRows(execute("SELECT " + function + "(value) FROM %s WHERE pk=0"), row(value));
    assertRows(execute("SELECT " + function + "([1, 2]) FROM %s WHERE pk=0"), row(value)); // Type error
}
```
The later example hits another bug when creating the function, where it says that the argument and the return type cannot be frozen. It's caused [here](https://github.com/dcapwell/cassandra/blob/CASSANDRA-18504/src/java/org/apache/cassandra/cql3/statements/schema/CreateFunctionStatement.java#L99) and [here](https://github.com/dcapwell/cassandra/blob/CASSANDRA-18504/src/java/org/apache/cassandra/cql3/statements/schema/CreateFunctionStatement.java#L103). I think we only need to add there an exception for vectors, the same way we do for tuples. Changing that lets the test pass the function creation, just to crash on the `WithList` type error.

Both examples for the token function and UDFs work if we use lists instead of vectors.

Reply:
:sigh: ill look into this... breaking `token` isn't good...

Reply:
thanks, added these failing tests for now, and will see about fixing...

Reply:
ok added the selector and these tests; they are both passing now.

I also updated random schema test to also check the token... just to make sure we do this for all type combos (which also would have hit this)